x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/classification/MulticlassConfusionMatrix.java

```diff
@@ -65,28 +67,36 @@ public String getName() {
     @Override
     public final List<AggregationBuilder> aggs(String actualField, String predictedField) {
         return List.of(
-            AggregationBuilders.composite(
-                AGG_NAME,
-                List.of(
-                    new TermsValuesSourceBuilder(ACTUAL_CLASS_FIELD).field(actualField),
-                    new TermsValuesSourceBuilder(PREDICTED_CLASS_FIELD).field(predictedField)
-                )
-            ).size(MAX_NUM_CLASSES * MAX_NUM_CLASSES)
+            AggregationBuilders.terms(AGGREGATE_BY_ACTUAL_CLASS)
+                .field(actualField)
+                .order(BucketOrder.key(true))
+                .size(MAX_NUM_CLASSES)
+                .subAggregation(
+                    AggregationBuilders.terms(AGGREGATE_BY_PREDICTED_CLASS)
+                        .field(predictedField)
+                        .order(BucketOrder.key(true))
+                        .size(MAX_NUM_CLASSES))
         );
     }
 
     @Override
     public Result evaluate(Aggregations aggs) {
-        CompositeAggregation agg = aggs.get(AGG_NAME);
         Map<String, Map<String, Long>> counts = new TreeMap<>();
-        for (CompositeAggregation.Bucket bucket : agg.getBuckets()) {
-            String actualClass = (String) bucket.getKey().get(ACTUAL_CLASS_FIELD);
-            String predictedClass = (String) bucket.getKey().get(PREDICTED_CLASS_FIELD);
-            Long docCount = bucket.getDocCount();
-            counts.putIfAbsent(actualClass, new TreeMap<>());
-            counts.get(actualClass).put(predictedClass, docCount);
+        Terms agg = aggs.get(AGGREGATE_BY_ACTUAL_CLASS);
+        long maxSumOfOtherDocCounts = agg.getSumOfOtherDocCounts();
+        for (Terms.Bucket bucket : agg.getBuckets()) {
+            String actualClass = bucket.getKeyAsString();
+            Map<String, Long> subCounts = new TreeMap<>();
+            counts.put(actualClass, subCounts);
+            Terms subAgg = bucket.getAggregations().get(AGGREGATE_BY_PREDICTED_CLASS);
+            maxSumOfOtherDocCounts = Math.max(maxSumOfOtherDocCounts, subAgg.getSumOfOtherDocCounts());
```

Comment:
I am a bit confused about this part and how it'd work. Here are my thoughts.

The set of actual classes may differ from that of the predicted classes. We're working with the first 1000 actual classes. For each of them, we're working with the first 1000 predicted classes for a given class.

I think it's fine in terms of the result matrix. It won't be a symmetric matrix, but I don't think it matters as we can still answer the question "how many times was class X classified as Y?".

However, when it comes to reporting the number of unhandled classes, I think what we do now may be confusing. There are 2 different counts at play. First, the count of unhandled actual classes which we get from the outer aggregation. Second, the count of unhandled predicted classes for each actual class we handle. I am not sure how helpful the max of all those is. Let's think a bit about this and discuss a solution.

Reply:
I like the general idea...

I think the natural way to implement this would be as follows:
1. Order the classes by frequency (unless there is some extrinsic notion of importance, i.e. user defined list),
2. Limit to 100 classes subject to the order defined in 1,
3. Introduce a new class "other" which is every class **not selected** in 2,
4. Report errors statistics for "actual is selected class prediction is other" and "actual is other prediction is selected type"

I'd probably omit the other vs other diagonal entry. Filling this in implies the classification is correct, where as of course we can't determine that by examining the actual classes.

Reply:
I think that would need 2 searches: the first to figure out the most frequent actual classes and the second to get the predicted classes after filtering out classes not in the above set.

We'll need to stretch the framework a bit to allow multiple searches but it might be good to do anyhow for paving the road for `auc_roc`, etc.

Reply:
Update:
I used `TypedChainTaskExecutor` to simplify the task chaining code.